### DISCUSSION TITLE
How to structure the _posts directory to include media and cover images for each post?

### DISCUSSION THREAD

**Discussion by Voxybuns** [jekyll/jekyll#8692] [Q&A]
I very recently begun working on a blog project using Jekyll, but I have a question and I hoped the community could provide some help or guidance!

I'd like to include images and videos inside my posts, as well as giving each of them a cover picture that could appear in the post template, but also as a thumbnail when generating the list of posts.

I was thinking using a structure similar to this:
```
_posts/
├─ yy-mm-dd-post-name/
│  ├─ yy-mm-dd-post-name.md
│  ├─ cover.png
│  ├─ my_image.png
│  ├─ another_image.png
├─ yy-mm-dd-post-name-2/
│  ├─ yy-mm-dd-post-name-2.md
│  ├─ cover.png
│  ├─ yet_another_image.png
│  ├─ sample_video.webm
├─ .../

```
Each post would have its own folder, which could also contain the aforementioned media as well as the cover picture. Would that work? Is there anything I should watch out for when trying to integrate any of those assets in my post and index templates? Thanks in advance for your help!

**Comment by chmaynard** [comment]
I'm using this technique. Apparently, static files inside a collection are not included in the ```site.static_files``` array.

**Comment by julien-blanchon** [comment]
I'm also curious ? I'm looking for the same kind of structure in order to make my markdown file render well on github or in my preview, so paths must be relative from the markdown file

**Comment by mohits** [comment]
I follow a relatively simple approach - there may be a better way.

in the `_posts` directory, I have the posts as usual.

```
2022-08-06-lessons-from-24-japan-for-project-managers.textile
2022-08-19-uninstalling-ruby-on-windows.textile
```

Then, completely separately, in the `images` directory, I have a directory called `posts` and then I add sub-directories as relevant. There is not a 1-to-1 relation between the posts and the folder in `images/posts` but it could be more like `images/posts/jekyll` or `images/posts/ruby/day1` and so on.

That way, the images are included by Jekyll when it collects them up. I just need to use the correct path to point to the images.

**Comment by hatamiarash7** [comment]
I have the same problem: [Include images in Jekyll collection](https://stackoverflow.com/questions/74039848/include-images-in-jekyll-collection)

**Comment by felipecrp** [comment]
It would.be a very useful feature to organize assets.

**Comment by Mohamed3nan** [comment]
It will make editing easier!

**Comment by RobertGauld** [comment]
Pleeeeease!

**Comment by ruttkowa** [comment]
I started to include the path to the images in the frontmatter / metadata on top like this
`asset_path: /assets/images/blog/2023-09-20/`

That way I can reference Images like `{{ page.asset_path }}image.jpg`

Bit cumbersome but works me - especially after writing a little script to generate this structure just like one would do with the HuGo static site generator (e.g. `hugo new content posts/my-first-post.md`)

**Comment by FlyersPh9** [comment]
I feel there has got to be a better way than this, but this is what I do currently:

```html
<img
  srcset="/assets/img/posts/{{ page.title | slugify }}/{{ include.url | replace: '.', '@2x.' }} 2x"
  src="/assets/img/posts/{{ page.title | slugify }}/{{ include.url }}"
  alt="{{ include.description }}"
/>

```

```ruby
{% include image.html url="my-photo.jpg" description="Lorem ipsum." %}
```

The reason I don't like this is that you need to create a post & a folder structure in a separate location:
- `/_posts/2023-12-23-my-post.md`
- `/assets/img/my-post/my-photo.jpg`

Ideally you would create one folder to place the post and related assets, then not need to reference that long path to make the image appear.

**Comment by yachting-earth** [comment]
jekyll-postfiles might solve what you are trying to accomplish https://nhoizey.github.io/jekyll-postfiles/

**Reply by PauloPhagula** [reply]
That has not been maintained for years and there are issues pilling on the issue tracker. Not something I'd recommend

**Comment by mikae1** [comment]
This is called [leaf bundles](https://www.engino.co.uk/form-elements/page-bundles/) in the Hugo world and I can highly recommend implementing it. It removes a lot of friction from posting.

**Comment by RDjarbeng** [comment]
Somebody please implement this so it's a default, this discussion has been running for 3 years😒. (since 2021)

**Comment by simonmysun** [comment]
It's bad that this is not supported. I'm using a rather ugly solution to keep the post media structured:

Directory layout (selected):

```
├── _posts
│   ├── 2024
│   │   ├── 2024-12-09-hello-world
│   │   │   ├── 2024-12-09-hello-world.md
│   │   │   └── image.md
```

The `image.md` is a picture encoded with Base64 and wrapped with Markdown image format, e.g. `![alt text](data:image/gif;base64,R0lGODlhwgFYAuYAAAAAAP///...)`. In the post, use `{% include_relative image.md %}` liquid syntax to add the image. 

Additionally, adjust the page where you list the posts to exclude the image (you may have to assign a property in each post. In my case, I'm filtering with `post.layout == 'post'`):

```liquid
{% for post in site.posts %}{% if post.layout == 'post' %}
...
{% endif %}{% endfor %}
```

Disadvantage:

- the generated HTML code becomes large and not very readable for human
  - On the other hand, it benefits from fewer requests and becomes faster. The increased traffic is not a concern because compression will reduce it and additional CPU load is hardly noticeable
- the images have to be encoded every time you add or edit them
  - Luckily you could still use the markdown preview to preview the images
- had never thought of SEO
  - Because I don't care

I hope this could have been done in a more elegant way.

**Reply by RDjarbeng** [reply]
Hope there was a more elegant way as well. I'm not sure if the devs are seeing all these comments in the thread. They could at least acknowledge receipt of the comments or something. 🙄

**Reply by ashmaroli** [reply]
@RDjarbeng, I have created an issue ticket at https://github.com/jekyll/jekyll/issues/9730 to *acknowledge notice of comments* in this thread. That said, it is uncertain if I will implement a solution towards this.

**Reply by RDjarbeng** [reply]
@ashmaroli thank you. Glad to see an acknowledgement👌. Hope for a resolution soon

**Comment by ashmaroli** [comment]
Hello everyone, I read all the comments in this thread and understand the desire to manage assets specific to blog posts.
But from the various comments here, it is clear that there is no consensus as to what is the best directory structure (regardless of Jekyll accepting it).

So, I have opened an issue ticket at #9730 to gauge *current interest* for this ability in the community and track development. Those interested, please leave upvote reactions (👍) on the opening post of the issue ticket.

**Comment by jeromevde** [comment]
I was able to achieve this by creating a custom plugin as follows:
Create a file named `copy_post_assets.rb` within the `_plugins` directory with the following content:
```
module Jekyll
    class CopyPostAssets < Generator
        safe true
        def generate(site)
            post_directory = File.join(site.source, '_posts') # Path to _posts directory
            if Dir.exist?(post_directory)
                Dir.glob(File.join(post_directory, '**', '*')) do |asset|
                    next unless File.file?(asset)             # Skip directories
                    next if File.extname(asset) == ".md"      # Skip .md files
                    relative_path = asset.sub(post_directory + '/', '') # Get the relative path
                    static_file = Jekyll::StaticFile.new(site, post_directory, "", relative_path)
                    site.static_files << static_file
                    # puts "Marked for future copy: #{asset} to #{File.join(site.dest, relative_path)}"
                end
                end
            else
                puts "No _posts directory found at: #{post_directory}"
            end
        end
    end
end
```
I think it could be improved by making it a bit more generic.
Let me know if it works for you:)

**Reply by RDjarbeng** [reply]
This might not work on GitHub pages for some deployments. I think plugins are disabled for GitHub pages

**Reply by jeromevde** [reply]
I was able to get it deployed on github pages actually by following these steps:
- add ``` gem "jekyll", "~> 4.2.2"``` in gemfile
- use a custom action workflow that allows to use a different jekyll version for deployment (the standard pages deployment seems to force you to use 3.9)

**Reply by ashmaroli** [reply]
Given that there is a perfectly viable alternative to classic GitHub Pages Deployment workflow, I really wish that Jekyll's active user base would adopt GitHub Actions, migrate to Jekyll 4.x and bid adieu to Jekyll 3.x.

**Reply by mrmanc** [reply]
It would be even better if GitHub gave Pages a bit of love and updated the stack to use Jekyll 4.x 🥺. I think the extra step of setting up actions for every new published site is a blocker for a decent proportion of Jekyll/GHP users, as they will often be using GH as a hosting platform rather than being experienced coders.

**Reply by RDjarbeng** [reply]
@jeromevde do you think doing this custom action workflow will work if you're using a Content Management System (CMS) for some of the sites. In CMS's like DecapCMS and SveltiaCMS the current way to work with assets is to include a path to the assets folder

For example
```
media_folder: "assets/images"  # Media files (images) will be stored here
public_folder: "/assets/images"
```

My fear is that using this custom script will either cause the flow of adding posts via the CMS to break or just be more painful to edit. 

IMO best thing would be to have it available by default in jekyll, then other tools would also be able to implement them in their future releases. Could decrease the likelihood of errors if enabled by default

**Reply by ashmaroli** [reply]
> It would be even better if GitHub gave Pages a bit of love and updated the stack to use Jekyll 4.x...

Technically, they cannot simply update the stack. Jekyll 4 has backwards-incompatible changes that could break existing sites deployed using *Classic GitHub Pages* experience. The practical approach is to provide an alternative route for those desiring control over build environment. And viola!, Hello, GitHub Actions!!

> I think the extra step of setting up actions for every new published site is a blocker for a decent proportion of Jekyll/GHP users...

While that is indeed a valid argument, my personal opinion is that benefits of Jekyll 4 (improved performance and reduced memory footprint) do not appear to be worth the hassle for users with small sites and those users who desired more from their SSG simply switched from Jekyll to another SSG.

**Reply by ashmaroli** [reply]
> best thing would be to have it available by default in jekyll...

The complications with native support for this feature:

- Posts and Documents from user-defined collections behave differently yet share few similar ones. (So, a practical approach would be to *pilot* the functionality for just posts initially and introduce the feature for other *documents* in a future release).
- `_posts` directory itself can be located inside any directory that do not start with an underscore (and certain other special characters) at any depth. So, users may expect similar behavior for placing assets relative to post document-files.
- `_posts` have documents with a _date-incorporated-filename_ whereas **`_drafts`** do not. Yet posts-specific assets have to be accessible to both drafts and posts alike.

**Reply by LorenzBung** [reply]
@jeromevde Thank you so much for this solution!!!
I am not deploying to github pages so this works perfectly fine for my circumstances.
There's an `end` too much in line 14 though.
